magento-research/pwa-studio
# Unified system for environment-based configuration

_This PR is a cherry-pick from #1198 to make that PR more digestible._ It needs to be merged before #1198.

## Description
Add tools in Buildpack to create, manage and validate configuration files for PWAs.

- Rename environment variables to fit a consistent UNDERSCORE_SEPARATED_NAMESPACE.
- Incorporate the brilliant new tool [camelspace](https://github.com/zetlen/camelspace) to abstract away the `process.env` and environment variables into configuration objects.
- Maintain a central `envVarDefinitions.json` in Buildpack for configuration across the project and all its tools:
    - Define the variables
    - Document the variables and their behavior
    - Validate according to simple types
    - Provide canonical defaults
- Create a `configureEnvironment` utility in Buildpack to replace the `validate-environment.js` script in Venia.
- Add functionality to announce deprecations of old environment variables, but continue to support them for some interval of time.

You can now translate easily between JS config objects and ENVIRONMENT_VARIABLE_OBJECTS, as well as drilling down to namespaces within them. Some tools (PWADevServer, upward-js) are already migrated to using these tools. Others are not quite yet.

## Related Issues
Closes #1116 

## Verification Steps
1. Run `yarn run build` to produce a working Buildpack CLI.
2. From repository root, run `node ./packages/pwa-buildpack/bin/buildpack create-env-file packages/venia-concept`
3. Observe the detailed logs and the created .env file.

## Proposed Labels for Change Type/Package




## Checklist:
